### PR TITLE
[X86][ISel] Fix logic for optimizing `movmsk(bitcast(shuffle(x)))`

### LINKED ISSUE #67287: Wrong code at -O2 on x86_64-linux_gnu since ddfee6d (recent regression)
Clang at -O2 produced the wrong code.

Bisected to ddfee6d0b6979fc6e61fa5ac7424096c358746fb, which was committed by @goldsteinn 

Compiler explorer: https://godbolt.org/z/rhfGPe4x5

```console
% cat a.c
int printf(const char *, ...);
long a, b, i;
char c, f;
int d;
short e;
int *g = &d;
unsigned h;
char j[1];
int main() {
  short *k = &e;
  for (; *g;)
    ;
  for (; c <= 30; c++) {
    if (b > 4)
      i = a;
    else
      i = b;
    (int)i + 6 || (*k = j != &f);
    *g ^= h;
  }
  printf("%d\n", e);
}
%
% clang a.c -O0 && ./a.out
0
% clang a.c -O2 && ./a.out
1
%
```

### PR DESCRIPTION
**Pull request by goldsteinn**
- [X86] Add tests for incorrectly optimizing out shuffle used in `movmsk`; PR67287
- [X86] Fix/improve logic for optimizing `movmsk(bitcast(shuffle(x)))`; PR67287

Fixes #67287.

### CODE REVIEW

File: llvm/test/CodeGen/X86/combine-ptest.ll [hunk 1]
```diff
@@ -265,7 +265,6 @@ define i32 @ptestz_v2i64_signbits(<2 x i64> %c, i32 %a, i32 %b) {
 ; SSE41-LABEL: ptestz_v2i64_signbits:
 ; SSE41:       # %bb.0:
 ; SSE41-NEXT:    movl %edi, %eax
-; SSE41-NEXT:    pshufd {{.*#+}} xmm0 = xmm0[1,1,3,3]
```
Comment (RKSimon) [inline llvm/test/CodeGen/X86/combine-ptest.ll:268]: This is definitely wrong as we've gone from just needing the signbits of a <2 x i64> to the signbits of a <4  x i32>

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 2]
```diff
@@ -45836,18 +45836,52 @@ static SDValue combineSetCCMOVMSK(SDValue EFLAGS, X86::CondCode &CC,
   // MOVMSK(SHUFFLE(X,u)) -> MOVMSK(X) iff every element is referenced.
   SmallVector<int, 32> ShuffleMask;
   SmallVector<SDValue, 2> ShuffleInputs;
+  SDValue BaseVec = peekThroughBitcasts(Vec);
   if (NumElts <= CmpBits &&
-      getTargetShuffleInputs(peekThroughBitcasts(Vec), ShuffleInputs,
-                             ShuffleMask, DAG) &&
+      getTargetShuffleInputs(BaseVec, ShuffleInputs, ShuffleMask, DAG) &&
       ShuffleInputs.size() == 1 && !isAnyZeroOrUndef(ShuffleMask) &&
       ShuffleInputs[0].getValueSizeInBits() == VecVT.getSizeInBits()) {
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:45843]: I think all we need to do is to add a scaleShuffleElements(ShuffleMask, NumElts, ScaledMask) check here to ensure that the shuffle mask can be scaled back to the original mask width:
```cpp
  // MOVMSK(SHUFFLE(X,u)) -> MOVMSK(X) iff every element is referenced.
  SmallVector<int, 32> ShuffleMask, ScaledMask;
  SmallVector<SDValue, 2> ShuffleInputs;
  if (NumElts <= CmpBits &&
      getTargetShuffleInputs(peekThroughBitcasts(Vec), ShuffleInputs,
                             ShuffleMask, DAG) &&
      ShuffleInputs.size() == 1 && !isAnyZeroOrUndef(ShuffleMask) &&
      ShuffleInputs[0].getValueSizeInBits() == VecVT.getSizeInBits() &&
      scaleShuffleElements(ShuffleMask, NumElts, ScaledMask)) {
    APInt DemandedElts = APInt::getZero(NumElts);
    for (int M : ScaledMask) {
      assert(0 <= M && M < (int)NumElts && "Bad unary shuffle index");
      DemandedElts.setBit(M);
    }
    if (DemandedElts.isAllOnes()) {
      SDLoc DL(EFLAGS);
      SDValue Result = DAG.getBitcast(VecVT, ShuffleInputs[0]);
      Result = DAG.getNode(X86ISD::MOVMSK, DL, MVT::i32, Result);
      Result =
          DAG.getZExtOrTrunc(Result, DL, EFLAGS.getOperand(0).getValueType());
      return DAG.getNode(X86ISD::CMP, DL, MVT::i32, Result,
                         EFLAGS.getOperand(1));
    }
  }
```

**Review by RKSimon — COMMENTED** [review]
nice catch!

File: llvm/test/CodeGen/X86/combine-ptest.ll (hunk 1 above)
Comment (goldsteinn) [reply llvm/test/CodeGen/X86/combine-ptest.ll:268]: Err, yeah I had flipped shuffle src/dst in the codes.

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 2 above)
Comment (goldsteinn) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:45843]: There are alot of ways to fix this. Posting V2 that I think is the most precise version. If you still have reservations can give this a try.

Comment (RKSimon) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:45843]: I'd prefer my proposal tbh

Comment (goldsteinn) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:45843]: The thing is its overly conservative. As long as all the "high" elements stay "high", we are fine. We don't need the proper words to stay together.

**Comment by RKSimon** [comment]
Have you any cases where all of this is necessary vs my smaller patch? This code was to handle a uncommon edge case in the first place (a movmsk variant of reduce(permute(x)) -> reduce(x)), and we're adding a lot of code to handle it. Ideally the SimplifyDemanded* calls will have simplified the shuffle anyhow if we're only demanding the sign bits .

**Comment by goldsteinn** [comment]
> Have you any cases where all of this is necessary vs my smaller patch? This code was to handle a uncommon edge case in the first place (a movmsk variant of reduce(permute(x)) -> reduce(x)), and we're adding a lot of code to handle it. Ideally the SimplifyDemanded* calls will have simplified the shuffle anyhow if we're only demanding the sign bits .

Okay, changed to your version + updated commit/PR message

**Review by RKSimon — APPROVED** [review]
LGTM - cheers!

**Comment by goldsteinn** [comment]
Pushed: https://github.com/llvm/llvm-project/commit/1684c65bc997a8ce0ecf96a493784fe39def75de

**Pull request closed without merging**